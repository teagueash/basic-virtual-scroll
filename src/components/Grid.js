import React from "react";
import PropTypes from "prop-types";
import "../styles/Grid.css";

/**
 * Grid component that is responsible for rendering desired data
 * @param {array} data        array of data passed from parent
 * @param {number} columnCount  variable representing the number of columns to render
 * @param {number} startIndex  variable indicating the current starting index of data items to display
 * @param {number} endIndex    variabe indicating the last index of data items to display
 * @param {object} style       configuration object containing element width, height, and offset values
 * @param {function} renderItem  function to render each data item in row/column
 */
const Grid = ({
  data,
  columnCount,
  startIndex,
  endIndex,
  style,
  renderItem
}) => {
  const { offset, elementHeight, elementWidth } = style;
  const batch = data.slice(startIndex, endIndex);

  return (
    <div className="grid">
      {batch.map(element => {
        const { index, background } = element;
        const row = Math.floor(index / columnCount);
        const column = index % columnCount;

        return (
          <div
            key={index}
            style={{
              top: `${row * (elementHeight + offset)}px`,
              left: `${column * (elementWidth + offset)}px`,
              width: `${elementWidth}px`,
              height: `${elementHeight}px`,
              background: `${background}`
            }}
            className="card"
          >
            {renderItem(element)}
          </div>
        );
      })}
    </div>
  );
};

Grid.propTypes = {
  data: PropTypes.array.isRequired,
  startIndex: PropTypes.number.isRequired,
  endIndex: PropTypes.number.isRequired,
  style: PropTypes.object.isRequired,
  renderItem: PropTypes.func.isRequired
};

export default Grid;
