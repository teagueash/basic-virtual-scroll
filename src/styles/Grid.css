@keyframes fade-in {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

.grid {
  animation: fade-in ease 1s forwards;
}

.card {
  position: absolute;
  text-align: center;
  box-sizing: border-box;
  background-color: #fff;
  border-radius: 4px;
  box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.2),
    0px 1px 1px 0px rgba(0, 0, 0, 0.14), 0px 2px 1px -1px rgba(0, 0, 0, 0.12);
}
