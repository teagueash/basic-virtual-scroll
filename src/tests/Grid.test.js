import React from "react";
import Grid from "../components/Grid.js";
import { shallow } from "enzyme";

// check that snapshot matches
it("will render Grid Component", () => {
  const mockRender = jest.fn();

  const mockProps = {
    data: [
      {
        index: 0,
        text: "lorem ipsum",
        background: "black"
      }
    ],
    range: 1,
    renderItem: mockRender,
    columnCount: 1,
    bufferCount: 0
  };
  const start = 0;
  const end = 1;
  const style = { offset: 5, elementWidth: 300, elementHeight: 300 };

  expect(
    shallow(
      <Grid {...mockProps} startIndex={start} endIndex={end} style={style} />
    )
  ).toMatchSnapshot();
});
