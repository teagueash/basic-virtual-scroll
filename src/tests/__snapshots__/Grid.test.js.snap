// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will render Grid Component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Grid
    bufferCount={0}
    columnCount={1}
    data={
      Array [
        Object {
          "background": "black",
          "index": 0,
          "text": "lorem ipsum",
        },
      ]
    }
    endIndex={1}
    range={1}
    renderItem={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "background": "black",
              "index": 0,
              "text": "lorem ipsum",
            },
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    startIndex={0}
    style={
      Object {
        "elementHeight": 300,
        "elementWidth": 300,
        "offset": 5,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="card"
          style={
            Object {
              "background": "black",
              "height": "300px",
              "left": "0px",
              "top": "0px",
              "width": "300px",
            }
          }
        />,
      ],
      "className": "grid",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "card",
          "style": Object {
            "background": "black",
            "height": "300px",
            "left": "0px",
            "top": "0px",
            "width": "300px",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="card"
            style={
              Object {
                "background": "black",
                "height": "300px",
                "left": "0px",
                "top": "0px",
                "width": "300px",
              }
            }
          />,
        ],
        "className": "grid",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "card",
            "style": Object {
              "background": "black",
              "height": "300px",
              "left": "0px",
              "top": "0px",
              "width": "300px",
            },
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
